### PR TITLE
feat(price-feeder): Convert remaining providers to the Websocket Controller

### PR DESCRIPTION
**Pull request by zarazan**
## Description

Converted the remaining providers over to use the Websocket Controller. For each provider I cross referenced the current API docs with how we currently handle their websocket and tuned it to maintain a stable connection.

Changes other than just converting providers to use the Controller:
- Coinbase: Added a ping loop to keep the connection alive longer
- Crypto.com: Changed the pricing info to parse into strings instead of float to correctly read prices
- Huobi: Add the HuobiSubscriptionResp to handle subscription response messages and prevent an erroneous error
- Added a second context to the WebsocketController to kill the old ping loop after a reconnect happens. Otherwise we risk running multiple ping loops for a single provider.
- Added an integration test that tests creating an active connection, receiving and parsing prices for each provider websocket we have implemented.

closes: https://github.com/umee-network/umee/issues/1507

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by adamewozniak** [comment]
woo! Checking this out @zarazan

File: price-feeder/oracle/provider/binance.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:74]: Add a comment here please

File: price-feeder/oracle/provider/binance.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:124]: Let's use a constant called something like `disabledPing` instead of `time.Duration(0)`, it's a little bit more obvious what you're doing

File: price-feeder/oracle/provider/binance.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:182]: Looks like `SubscribeCurrencyPairs` is a no-op on every provider now, should we remove it from the provider interface & stop using it altogether? 

Also, if we're keeping this in for reverse-compatibility, please update the function comments instead
```golang
// SubscribeCurrencyPairs currently performs a no-op; this is handled by the websocket controller.
func (p *BinanceProvider) SubscribeCurrencyPairs(cps 
```

File: price-feeder/oracle/provider/binance.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:241]:
```suggestion
```

File: price-feeder/oracle/provider/bitget.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/bitget.go:229]: Guessing this was the source of many logs

File: price-feeder/oracle/provider/crypto.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/crypto.go:339]: Do we not still have to convert this from seconds -> milliseconds? 

```suggestion
		SecondsToMilli(candlePair.Timestamp),
```

File: price-feeder/oracle/provider/gate.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/gate.go:21]: nice. still need to build something that tells us when these apis update

File: price-feeder/oracle/provider/huobi.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/huobi.go:81]: Comment here as well pls

File: price-feeder/oracle/provider/websocket_controller.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/websocket_controller.go:259]: Do we need this function?

File: price-feeder/oracle/provider/websocket_controller.go
Comment (adamewozniak) [inline price-feeder/oracle/provider/websocket_controller.go:261]: Might want to add a comment

```suggestion
// pingHandler is used to respond to pings from xyz
func (wsc *WebsocketController) pingHandler(appData string) error {
```

**Review by adamewozniak — COMMENTED** [review]
a few comments but this looks good so far! good job @zarazan !

File: price-feeder/oracle/provider/binance.go
Comment (zarazan) [reply price-feeder/oracle/provider/binance.go]: https://github.com/umee-network/umee/pull/1554/commits/ff5511dfe2f61792f9d96a2cc77b14ce8df2844a

File: price-feeder/oracle/provider/binance.go
Comment (zarazan) [reply price-feeder/oracle/provider/binance.go:75]: https://github.com/umee-network/umee/pull/1554/commits/ff5511dfe2f61792f9d96a2cc77b14ce8df2844a

File: price-feeder/oracle/provider/huobi.go
Comment (zarazan) [reply price-feeder/oracle/provider/huobi.go:82]: https://github.com/umee-network/umee/pull/1554/commits/ff5511dfe2f61792f9d96a2cc77b14ce8df2844a

File: price-feeder/oracle/provider/binance.go
Comment (zarazan) [reply price-feeder/oracle/provider/binance.go]: zeroed it https://github.com/umee-network/umee/pull/1554/commits/ab4d240ad9caf2dfe37ca00e5db36906d9016d88

File: price-feeder/oracle/provider/binance.go
Comment (zarazan) [reply price-feeder/oracle/provider/binance.go]: agreed. https://github.com/umee-network/umee/pull/1554/commits/ef919f977998d74fd192688a008d607876dec610

File: price-feeder/oracle/provider/crypto.go
Comment (zarazan) [reply price-feeder/oracle/provider/crypto.go]: good catch, for some reason I thought NewCandlePrice did that. https://github.com/umee-network/umee/pull/1554/commits/ef919f977998d74fd192688a008d607876dec610

File: price-feeder/oracle/provider/websocket_controller.go
Comment (zarazan) [reply price-feeder/oracle/provider/websocket_controller.go:260]: No, we don't. It's just nice to have there for when you need to test a provider and see if they pong you back. I think it can be removed.

Comment (zarazan) [reply price-feeder/oracle/provider/websocket_controller.go]: https://github.com/umee-network/umee/pull/1554/commits/166014a0f416f06f581659b8dea14a521c7e3d42

File: price-feeder/oracle/provider/websocket_controller.go
Comment (zarazan) [reply price-feeder/oracle/provider/websocket_controller.go:255]: https://github.com/umee-network/umee/pull/1554/commits/166014a0f416f06f581659b8dea14a521c7e3d42

File: price-feeder/oracle/provider/binance.go
Comment (adamewozniak) [reply price-feeder/oracle/provider/binance.go]: niceu. this will change the PF / peggo integration. 

cc: @RafilxTenfen

**Review by adamewozniak — APPROVED** [review]
looks good! just need tests to pass.

would like another review, maybe from @rbajollari / @RafilxTenfen before we merge this in

**Comment by adamewozniak** [comment]
oh, and changelog please!

**Comment by RafilxTenfen** [comment]
With this implementation, peggo cannot update the price-feeder version anymore, basically peggo does not know which prices will have to query when instantiating new providers. The new tickers might appear when someone tries to cross a different token

Or maybe every time a new price comes, we would need to kill the providers and instantiate again

**Comment by RafilxTenfen** [comment]
this is so much cleaner than it was before, but thinking of price-feeder as a package for others, would be good to have an func to subscribe new pairs as needed with the provider running

**Comment by adamewozniak** [comment]
> this is so much cleaner than it was before, but thinking of price-feeder as a package for others, would be good to have an func to subscribe new pairs as needed with the provider running

cc @zarazan

File: price-feeder/oracle/provider/osmosisv2_test.go
Comment (zarazan) [reply price-feeder/oracle/provider/osmosisv2_test.go:116]: It didn't fail so I didn't notice it I guess. Good call out, I'm adding this function back in anyway.

**Comment by adamewozniak** [comment]
@zarazan please lmk once rafael's concerns are addressed 🚀

**Review by adamewozniak — APPROVED** [review]
this looks good to me, might want another review from @RafilxTenfen

**Comment by toteki** [comment]
Do these need to be backported for price feeder v1 branch or is it all v2 stuff?

**Pull request merged**
